tour_model_eval/evaluation_pipeline.py

```diff
@@ -0,0 +1,173 @@
+        # updated with the labels after two rounds of clustering.
+        new_labels = first_labels.copy()
+        first_label_set = list(set(first_labels))
+        track = get_track(bins, first_labels)
+        # collect tuning scores and parameters
+        tune_score = {}
+        for dist_pct in np.arange(0.15, 0.6, 0.02):
+            for low in range(250, 600):
+                percentage_second, homo_second = second_round(first_label_set, first_labels, bin_trips, filter_trips,
+                                                              low, dist_pct,
+                                                              sim, new_labels, track)
+
+                curr_score = gs.get_score(homo_second, percentage_second)
+                if curr_score not in tune_score:
+                    tune_score[curr_score] = (low, dist_pct, homo_second, percentage_second)
+
+        best_score = max(tune_score)
+        sel_tradeoffs = tune_score[best_score][0:2]
+    else:
+        sel_tradeoffs = (0,0)
+
+    return sel_tradeoffs
+
+
+def test(data,radius,low,dist_pct):
+    sim, bins, bin_trips, filter_trips = first_round(data, radius)
+    # it is possible that we don't have common trips for tuning or testing
+    # bins contain common trips indices
+    if len(bins) is not 0:
+        gs.compare_trip_orders(bins, bin_trips, filter_trips)
+        first_labels = get_first_label(bins)
+        # new_labels temporary stores the labels from the first round, but later the labels in new_labels will be
+        # updated with the labels after two rounds of clustering.
+        new_labels = first_labels.copy()
+        first_label_set = list(set(first_labels))
+        track = get_track(bins, first_labels)
+        # get request percentage for the subset for the first round
+        percentage_first = grp.get_req_pct(new_labels, track, filter_trips, sim)
+        # get homogeneity score for the subset for the first round
+        homo_first = gs.score(bin_trips, first_labels)
+        percentage_second, homo_second = second_round(first_label_set, first_labels, bin_trips, filter_trips, low,
+                                                      dist_pct, sim, new_labels, track)
+    else:
+        percentage_first = 1
+        homo_first = 1
+        percentage_second = 1
+        homo_second = 1
+    scores = gs.get_score(homo_second, percentage_second)
+    return homo_first,percentage_first,homo_second,percentage_second,scores
+
+
+def main(uuid = None):
+    user = uuid
+    radius = 100
+    trips = preprocess.read_data(user)
+    filter_trips = preprocess.filter_data(trips, radius)
+    tune_idx, test_idx = preprocess.split_data(filter_trips)
+    tune_data = preprocess.get_subdata(filter_trips, test_idx)
+    test_data = preprocess.get_subdata(filter_trips, tune_idx)
+    pct_collect_first = []
+    homo_collect_first = []
+    pct_collect_second = []
+    homo_collect_second = []
+    coll_score = []
+    coll_tradeoffs = []
+
```

Comment:
Again, can this be moved into a list of results or even a dataframe instead of multiple lists of results?

Reply:
I have changed to use a dataframe to collect all clustering result. These lists are deleted.